cpp/unreal_plugins/SimulationController/Source/SimulationController/UrdfBotAgent.cpp

```diff
@@ -0,0 +1,179 @@
+    urdf_bot_pawn_ = nullptr;
+}
+
+void UrdfBotAgent::findObjectReferences(UWorld* world)
+{
+    auto step_info_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.URDFBOT_AGENT.STEP_INFO_COMPONENTS");
+}
+
+void UrdfBotAgent::cleanUpObjectReferences()
+{
+    auto step_info_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.URDFBOT_AGENT.STEP_INFO_COMPONENTS");
+}
+
+std::map<std::string, Box> UrdfBotAgent::getActionSpace() const
+{
+    std::map<std::string, Box> action_space;
+    auto action_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.URDFBOT_AGENT.ACTION_COMPONENTS");
+
+    if (Std::contains(action_components, "joint")) {
+        for (auto& joint_pair : urdf_bot_pawn_->urdf_robot_component_->joint_components_) {
+            if (joint_pair.second->control_type_ != UrdfJointControlType::Invalid) {
+                Box box;
+                box.low_ = std::numeric_limits<float>::lowest();
+                box.high_ = std::numeric_limits<float>::max();
+                box.shape_ = {1};
+                box.datatype_ = DataType::Float32;
+                action_space["joint." + joint_pair.first] = std::move(box);
+            }
+        }
+    }
+
+    return action_space;
+}
+
+std::map<std::string, Box> UrdfBotAgent::getObservationSpace() const
+{
+    std::map<std::string, Box> observation_space;
+
+    auto observation_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.URDFBOT_AGENT.OBSERVATION_COMPONENTS");
+
+    if (Std::contains(observation_components, "link_state")) {
+        for (auto& link_pair : urdf_bot_pawn_->urdf_robot_component_->link_components_) {
+            Box box;
+            box.low_ = std::numeric_limits<float>::lowest();
+            box.high_ = std::numeric_limits<float>::max();
+            box.shape_ = {6};
+            box.datatype_ = DataType::Float32;
+            observation_space["link_state." + link_pair.first] = std::move(box);
+        }
+    }
+
+    std::map<std::string, Box> camera_sensor_observation_space = camera_sensor_->getObservationSpace(observation_components);
+    for (auto& camera_sensor_observation_space_component : camera_sensor_observation_space) {
+        observation_space[camera_sensor_observation_space_component.first] = std::move(camera_sensor_observation_space_component.second);
+    }
+
+    return observation_space;
+}
+
+std::map<std::string, Box> UrdfBotAgent::getStepInfoSpace() const
+{
+    std::map<std::string, Box> step_info_space;
+
+    return step_info_space;
+}
+
+void UrdfBotAgent::applyAction(const std::map<std::string, std::vector<uint8_t>>& actions)
+{
+    auto action_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.URDFBOT_AGENT.ACTION_COMPONENTS");
+    std::map<std::string, float> joint_actions;
+
+    for (auto& action : actions) {
+        if (Std::contains(action_components, "joint") && action.first.find("joint.") == 0) {
```

Comment:
There is already a string processing section in `Std::`. Let's replace this right-hand if statement with `Std::startsWith(action.first, "joint.")`.

(You'll need to create the `startsWith` function, but I think what you're doing here will be a recurring pattern for named actions, so it'll be nice to have a more readable variant of `str.find("joint.") == 0`.)

I think it would be cleaner if the internal names in `urdf_robot_component_` were named `joint.whatever` instead of `whatever`. That way, you wouldn't need to parse out a new `joint_name`. This would be slightly more consistent with how `CameraSensor` works.

Reply:
moved `getObservation` and `applyAction` to `UrdfRobotComponent` while keep `getSpace` in `UrdfAgent`